modelzoo/Unet.py

```diff
@@ -0,0 +1,62 @@
+
+    net['contr_3_1'] = ConvLayer(net['pool2'], base_n_filters*4, 3, nonlinearity=nonlinearity, pad=pad)
+    net['contr_3_2'] = ConvLayer(net['contr_3_1'], base_n_filters*4, 3, nonlinearity=nonlinearity, pad=pad)
+    net['pool3'] = Pool2DLayer(net['contr_3_2'], 2)
+
+    net['contr_4_1'] = ConvLayer(net['pool3'], base_n_filters*8, 3, nonlinearity=nonlinearity, pad=pad)
+    net['contr_4_2'] = ConvLayer(net['contr_4_1'], base_n_filters*8, 3, nonlinearity=nonlinearity, pad=pad)
+    l = net['pool4'] = Pool2DLayer(net['contr_4_2'], 2)
+    # the paper does not really describe where and how dropout is added. Feel free to try more options
+    if do_dropout:
+        l = DropoutLayer(l, p=0.5)
+
+    net['encode_1'] = ConvLayer(l, base_n_filters*16, 3, nonlinearity=nonlinearity, pad=pad)
+    net['encode_2'] = ConvLayer(net['encode_1'], base_n_filters*16, 3, nonlinearity=nonlinearity, pad=pad)
+    net['deconv1'] = Deconv2DLayer(net['encode_2'], base_n_filters*8, 2, 2)
+
+    net['concat1'] = ConcatLayer([net['deconv1'], net['contr_4_2']], cropping=(None, None, "center", "center"))
+    net['expand_1_1'] = ConvLayer(net['concat1'], base_n_filters*8, 3, nonlinearity=nonlinearity, pad=pad)
+    net['expand_1_2'] = ConvLayer(net['expand_1_1'], base_n_filters*8, 3, nonlinearity=nonlinearity, pad=pad)
+    net['deconv2'] = Deconv2DLayer(net['expand_1_2'], base_n_filters*4, 2, 2)
+
+    net['concat2'] = ConcatLayer([net['deconv2'], net['contr_3_2']], cropping=(None, None, "center", "center"))
+    net['expand_2_1'] = ConvLayer(net['concat2'], base_n_filters*4, 3, nonlinearity=nonlinearity, pad=pad)
+    net['expand_2_2'] = ConvLayer(net['expand_2_1'], base_n_filters*4, 3, nonlinearity=nonlinearity, pad=pad)
+    net['deconv3'] = Deconv2DLayer(net['expand_2_2'], base_n_filters*2, 2, 2)
+
+    net['concat3'] = ConcatLayer([net['deconv3'], net['contr_2_2']], cropping=(None, None, "center", "center"))
+    net['expand_3_1'] = ConvLayer(net['concat3'], base_n_filters*2, 3, nonlinearity=nonlinearity, pad=pad)
+    net['expand_3_2'] = ConvLayer(net['expand_3_1'], base_n_filters*2, 3, nonlinearity=nonlinearity, pad=pad)
+    net['deconv4'] = Deconv2DLayer(net['expand_3_2'], base_n_filters, 2, 2)
+
+    net['concat4'] = ConcatLayer([net['deconv4'], net['contr_1_2']], cropping=(None, None, "center", "center"))
+    net['expand_4_1'] = ConvLayer(net['concat4'], base_n_filters, 3, nonlinearity=nonlinearity, pad=pad)
+    net['expand_4_2'] = ConvLayer(net['expand_4_1'], base_n_filters, 3, nonlinearity=nonlinearity, pad=pad)
+
+    net['segLayer'] = ConvLayer(net['expand_4_2'], num_output_classes, 1, nonlinearity=None)
+    net['dimshuffle'] = DimshuffleLayer(net['segLayer'], (1, 0, 2, 3))
+    net['reshapeSeg'] = ReshapeLayer(net['dimshuffle'], (num_output_classes, -1))
+    net['dimshuffle2'] = DimshuffleLayer(net['reshapeSeg'], (1, 0))
+    net['output'] = NonlinearityLayer(net['dimshuffle2'], nonlinearity=lasagne.nonlinearities.softmax)
```

Comment:
There is no need to reverse the dimshuffle and reshape operations. I think what you are looking for is net['segLayer']. It has the exact shape you described in your comment: (batchsize, classes, height, width). I will edit the names so that they are more clear

Lasagne/Lasagne#626 alone would not help much because the shape after the new softmax is still incompatible with the categorical crossentropy implementation. 
I tried using
def softmax(x):
    e_x = theano.tensor.exp(x - x.max(axis=1, keepdims=True))
    return e_x / e_x.sum(axis=1, keepdims=True)
But the 4D output of the softmax did not work with the lasagne categorical crossentropy imlementation. I had a quick glance at the implementation: It did not look like changing it would be easy which is why I stuck with dimshuffle + reshape